camel-knative/camel-knative-api/src/main/java/org/apache/camel/component/knative/spi/KnativeEnvironment.java

```diff
@@ -53,24 +53,24 @@ public KnativeEnvironment(
         return services.stream();
     }
 
-    public Optional<KnativeServiceDefinition> lookupService(Knative.Type type, String name, String... aliases) {
+    public Optional<KnativeServiceDefinition> lookupService(Knative.Kind endpointType, Knative.Type type, String name, String... aliases) {
         return Stream.concat(Stream.of(name), Stream.of(aliases))
             .sequential()
-            .map(n -> lookup(type, n))
+            .map(n -> lookup(endpointType, type, n))
             .filter(Optional::isPresent)
             .map(Optional::get)
             .findFirst();
     }
 
-    public KnativeServiceDefinition mandatoryLookupService(Knative.Type type, String name) {
-        return lookupService(type, name).orElseThrow(
+    public KnativeServiceDefinition mandatoryLookupService(Knative.Kind endpointType, Knative.Type type, String name) {
```

Comment:
In case you want to add support for 4 fields lookup in this PR, `apiVersion` and (Kubernetes) `kind` are optional, so you should filter them only when explicitly set on the endpoint.

Reply:
Yes, it's possible that one uses `kind=Channel` and it can mean `messaging.knative.dev/v1alpha1` or `eventing.knative.dev/v1alpha1` (the operator will resolve the right one based on what is present in the cluster). Or in the future there can be multiple possible API versions of the same group and a Camel user will just declare the kind.